### PR TITLE
Articles API: create and update

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description

This PR accomplishes two main things:

* moves the editor to use regular controllers instead of the articles API controllers
* reworks the articles API to use the API key and adds a slightly better interface to the outside world

A few points of discussion:

* should the description be part of the public API? Currently it's either taken from the front matter or from the article's body:

https://github.com/thepracticaldev/dev.to/blob/master/app/models/article.rb#L491

* the collection seems to be reset if there's a title in the front matter, is this correct or a bug?:

https://github.com/thepracticaldev/dev.to/blob/master/app/models/article.rb#L492

* should we send a notification if the previously unpublished article becomes published during the update?

see https://github.com/thepracticaldev/dev.to/blob/master/app/services/article_creation_service.rb#L20

## Related Tickets & Documents

Refers to #911 
Closes #2040

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

File: app/views/api/v0/articles/show.json.jbuilder [hunk 1]
```diff
@@ -7,6 +7,7 @@ json.published_at       @article.published_at
 json.readable_publish_date @article.readable_publish_date
 json.social_image       article_social_image_url(@article)
 json.tag_list           @article.cached_tag_list
+json.tags               @article.cached_tag_list.split(", ")
```
Comment (Turnerj) [inline app/views/api/v0/articles/show.json.jbuilder]: Would it be better using `@article.cached_tag_list_array` instead like the `index.json.jbuilder`?
https://github.com/thepracticaldev/dev.to/blob/d66da649dd2978e5bcda868ed5d4cf0df429c54c/app/views/api/v0/articles/index.json.jbuilder#L8

**Comment by benhalpern** [comment]
> should the description be part of the public API? Currently it's either taken from the front matter or from the article's body:

Yeah, I think it should. 

> the collection seems to be reset if there's a title in the front matter, is this correct or a bug?:

Basically if there is _any_ frontmatter, the whole API reads the frontmatter instead of the given params. So if I add a `title` param but also send over a `body_markdown` which contains the title, that will be read and the params will not be. This line ensures that the `collection_id` is set to `nil` if there is frontmatter, because otherwise if `series` is included in the frontmatter and then removed, the update wouldn't know to remove it. This is also essentially done above with tags.

I'm sure this could be cleaner but that's the logic. 

> should we send a notification if the previously unpublished article becomes published during the update?

Yeah. It occurs to me now that we were doing this in the `articles_controller` but failed to maintain this functionality when switching over to the new controller.

***

Overall this PR looks great, and the existing functionality looks to be preserved well, so I'm going to merge this shortly.

FYI: The "HTML" path of the creation/updating of the old controller is no longer needed and is dead code that should have been removed along with the change. And I'm seeing some bugs in terms of behavior which was never included in "v2" style.

You're essentially putting the existing "out of your way" with this PR, so I'll get to working on fixing those hiccups. We should eventually converge on using `Updater` everywhere.

**Comment by rhymes** [comment]
Hi @benhalpern,

> Yeah, I think it should.

But how do we get it from the incoming JSON to the DB without changing the internal logic? If we accept the description from the API now it's going to get overwritten by the `evaluate_front_matter` anyway.

> Yeah. It occurs to me now that we were doing this in the articles_controller but failed to maintain this functionality when switching over to the new controller.

Got it, the drawback that occurred to me is that it would send a notification each time. It's likely a rare corner case but if I were to flip unpublished to published to unpublished to published I'd be sending two notifications, one each time the article gets updated!

> Overall this PR looks great, and the existing functionality looks to be preserved well, so I'm going to merge this shortly.

I did the following for both create and update: wrote and ran a test, moved over the logic, swapped the endpoint, re-ran the test.

> FYI: The "HTML" path of the creation/updating of the old controller is no longer needed and is dead code that should have been removed along with the change.

Oooh that's why, I'm all in favor of removing dead code 👍 

> and I'm seeing some bugs in terms of behavior which was never included in "v2" style

you mean both in the API and the regular controllers?

> You're essentially putting the existing "out of your way" with this PR, so I'll get to working on fixing those hiccups. We should eventually converge on using Updater everywhere.

cool!



Anyway thanks for the review, the logic you explained makes a lot of sense! I'll see if we can simplify anything after your changes/fixes

**Comment by benhalpern** [comment]
> But how do we get it from the incoming JSON to the DB without changing the internal logic? If we accept the description from the API now it's going to get overwritten by the evaluate_front_matter anyway.

Same way we do with other frontmatter elements, accept it if it's there as a `description` param. But if you are sending things via frontmatter, only look at `body_markdown`.

> Got it, the drawback that occurred to me is that it would send a notification each time. It's likely a rare corner case but if I were to flip unpublished to published to unpublished to published I'd be sending two notifications, one each time the article gets published!

I believe the way we are currently dealing with this is by checking whether `published_at` changed. `published` will always changed, but `published_at` is a date that would only change the first time.

> you mean both in the API and the regular controllers?

Yeah, I believe we've had some logical inconsistencies the whole time. Certain functionality _only_ fires in the most legacy version of the `ArticlesController`. So v2 was always missing a bit of functionality in terms of updating the `edited_at` and sending the notification.

**Pull request merged**